src/modules/lost/functions.c

```diff
@@ -0,0 +1,551 @@
+			if(idhdr.len == 0) {
+				LM_ERR("device id not found\n");
+				goto err;
+			}
+		}
+		did.s = idhdr.s;
+		did.len = idhdr.len;
+	}
+	LM_INFO("id - [%.*s]\n", did.len, did.s);
+
+	/* check if connection exists */
+	if(httpapi.http_connection_exists(&con) == 0) {
+		LM_ERR("connection: [%s] does not exist\n", con.s);
+		goto err;
+	}
+
+	/* assemble locationRequest */
+	que.s = lost_held_location_request(did.s, &que.len);
+	/* free memory */
+	if(idhdr.s) {
+		pkg_free(idhdr.s);
+		idhdr.len = 0;
+		did.s = NULL;
+		did.len = 0;
+	}
+	if(!que.s) {
+		LM_ERR("held request document error\n");
+		goto err;
+	}
+
+	LM_DBG("held location request: \n[%s]\n", que.s);
+
+	/* send locationRequest to location server - HTTP POST */
+	httpapi.http_connect(_m, &con, NULL, &res, mtheld, &que);
+	/* free memory */
+	pkg_free(que.s);
+	que.len = 0;
+	/* read and parse the returned xml */
+	doc = xmlReadMemory(res.s, res.len, 0, NULL,
+			XML_PARSE_RECOVER | XML_PARSE_NOBLANKS | XML_PARSE_NONET
+					| XML_PARSE_NOCDATA);
+	if(!doc) {
+		LM_WARN("invalid xml document: \n[%.*s]\n", res.len, res.s);
+		doc = xmlRecoverMemory(res.s, res.len);
+		if(!doc) {
+			LM_ERR("xml document recovery failed on: \n[%.*s]\n", res.len,
+					res.s);
+			goto err;
+		}
+		LM_DBG("xml document recovered\n");
+	}
+	root = xmlDocGetRootElement(doc);
+	if(!root) {
+		LM_ERR("empty xml document\n");
+		goto err;
+	}
+	/* check the root element, shall be locationResponse, or errors */
+	if(!xmlStrcmp(root->name, (const xmlChar *)"locationResponse")) {
+
+		LM_DBG("HELD location response: \n[%.*s]\n", res.len, res.s);
+
+		/* get the locationUri element */
+		geo.s = lost_get_content(root, (char *)"locationURI", &geo.len);
+		if(!geo.s) {
+			LM_ERR("no locationURI element found\n");
+			goto err;
+		}
+	} else if(!xmlStrcmp(root->name, (const xmlChar *)"error")) {
+
+		LM_DBG("HELD error response: \n[%.*s]\n", res.len, res.s);
+
+		/* get the error patterm */
+		err.s = lost_get_property(root, (char *)"code", &err.len);
+		if(!err.s) {
+			LM_ERR("error - code property not found: \n[%.*s]\n", res.len,
+					res.s);
+			goto err;
+		}
+		LM_WARN("locationRequest error response: [%.*s]\n", err.len, err.s);
+	} else {
+		LM_ERR("root element is not valid: \n[%.*s]\n", res.len, res.s);
+		goto err;
+	}
+	xmlFreeDoc(doc);
+
+	/* set writeable pvars */
+	pvpidf.rs = res;
+	pvpidf.rs.s = res.s;
+	pvpidf.rs.len = res.len;
+
+	pvpidf.flags = PV_VAL_STR;
+	pspidf = (pv_spec_t *)_pidf;
+	pspidf->setf(_m, &pspidf->pvp, (int)EQ_T, &pvpidf);
+
+	pvurl.rs = geo;
+	pvurl.rs.s = geo.s;
+	pvurl.rs.len = geo.len;
+
+	pvurl.flags = PV_VAL_STR;
+	psurl = (pv_spec_t *)_url;
+	psurl->setf(_m, &psurl->pvp, (int)EQ_T, &pvurl);
+
+	pverr.rs = err;
+	pverr.rs.s = err.s;
+	pverr.rs.len = err.len;
+
+	pverr.flags = PV_VAL_STR;
+	pserr = (pv_spec_t *)_err;
+	pserr->setf(_m, &pserr->pvp, (int)EQ_T, &pverr);
+
+	return (err.len > 0) ? LOST_SERVER_ERROR : LOST_SUCCESS;
+
+err:

```

Comment:
You are basically leaking memory here, in case one of the memory allocation success initial and a later one (or another library call) is jumping into err. It is probably ncessary to add some (if XXX.s) { pkg_free(XXX.s) } calls here.